### PR TITLE
Dec 21 2021

### PR DESCRIPTION
**Pull request by Rahul4321Kumar**
This PR contains the following:-
1. Added two app (i.e user_app, Enquiry)
2. Added models of user and Enquiry
3. used abstractbaseuser to create user_app model

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,6 @@
+*.log
+*.pot
+*.pyc
+__pycache__/
+local_settings.py
+db.sqlite3
```
Comment (ksilawat-deqode) [inline .gitignore]: No newline at the end

File: EMS/EMS/settings.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    'django.contrib.admin',
+    'django.contrib.auth',
+    'django.contrib.contenttypes',
+    'django.contrib.sessions',
+    'django.contrib.messages',
+    'django.contrib.staticfiles',
+    'Enquiry',
+    'user_app',
+    'rest_framework',
+]
```
Comment (ksilawat-deqode) [inline EMS/EMS/settings.py]: Here apps can be separated like,
```suggestion
INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    
    #integrated apps
    'rest_framework',
    
    #own apps
    'enquiries',
    'users',
]
```

File: EMS/Enquiry/admin.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+from django.contrib import admin
+from .models import Enquiry
+# Register your models here.
+admin.site.register(Enquiry)
```
Comment (ksilawat-deqode) [inline EMS/Enquiry/admin.py]: Here 2 lines break should be present and then only you should start your implementations
```suggestion
from django.contrib import admin
from .models import Enquiry


admin.site.register(Enquiry)
```

File: EMS/Enquiry/admin.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
+from django.contrib import admin
+from .models import Enquiry
```
Comment (ksilawat-deqode) [inline EMS/Enquiry/admin.py]: Always use absolute imports and not relative imports
```suggestion
from enquiries.models import Enquiry
```

File: EMS/Enquiry/models.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from django.db import models
+from django.conf import settings
+# Create your models here.
+class Enquiry(models.Model):
```
Comment (ksilawat-deqode) [inline EMS/Enquiry/models.py]:
```suggestion
from django.db import models
from django.conf import settings


class Enquiry(models.Model):
```

File: EMS/Enquiry/models.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+    message = models.TextField(max_length=200)
+    active = models.BooleanField()
```
Comment (ksilawat-deqode) [inline EMS/Enquiry/models.py]: You should add a default value for this `active` field.

File: EMS/Enquiry/models.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now_add=True)
```
Comment (ksilawat-deqode) [inline EMS/Enquiry/models.py]: `created_at` is for at what time instance is created
 `updated_at` is for at what time instance is updated(or we can last modify)
But with your implementation, both fields are doing the same. Please read about `auto_now` & `auto_now_add`

File: EMS/user_app/admin.py [hunk 8]
```diff
@@ -0,0 +1,4 @@
+from django.contrib import admin
+from .models import User
+# Register your models here.
+admin.site.register(User)
```
Comment (ksilawat-deqode) [inline EMS/user_app/admin.py]:
```suggestion
from django.contrib import admin
from users.models import User


admin.site.register(User)
```

File: EMS/user_app/managers.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
+from django.contrib.auth.base_user import BaseUserManager
```
Comment (ksilawat-deqode) [inline EMS/user_app/managers.py]: No need to create an extra file for only model managers, model managers reside inside `models.py`

File: EMS/user_app/models.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+from django.contrib.auth.base_user import AbstractBaseUser
+from django.utils.translation import ugettext_lazy as _
+
+from .managers import UserManager
+
+# Create your models here.
+class User(AbstractBaseUser, PermissionsMixin):
+    GENDER_CHOICES = (
+        (u'M', u'Male'),
+        (u'F', u'Female'),
```
Comment (ksilawat-deqode) [inline EMS/user_app/models.py]: Any specific reason for not using normal strings here?

**Review by ksilawat-deqode — CHANGES_REQUESTED** [review]
Still needs improvement!
It's already mentioned in the doc for the names of the apps -> `users` & `enquiries`
Still you're having `user_app` in place of `users` & `Enquiry` in place of `enquiries`

Django apps names always follows snakecase!

Please focus on naming conventions and Review comments as well.

It doesn't contain any `requirements.txt`, please add it

File: EMS/enquiries/admin.py [hunk 11]
```diff
@@ -0,0 +1,5 @@
+from django.contrib import admin
+from enquiries.models import Enquiry
+# Register your models here.
+
+admin.site.register(Enquiry)
```
Comment (ksilawat-deqode) [inline EMS/enquiries/admin.py]: No new line at the end

File: EMS/enquiries/models.py [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+from django.conf import settings
+# Create your models here.
+
+class Enquiry(models.Model):
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+    message = models.TextField(max_length=200)
+    active = models.BooleanField(default=True)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    resolved = models.BooleanField()
```
Comment (ksilawat-deqode) [inline EMS/enquiries/models.py]: Add `default=False`.

File: EMS/users/admin.py [hunk 13]
```diff
@@ -0,0 +1,5 @@
+from django.contrib import admin
+from users.models import User
+# Register your models here.
+
+admin.site.register(User)
```
Comment (ksilawat-deqode) [inline EMS/users/admin.py]: No newline at the end

File: EMS/users/models.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
+from django.db import models
+from django.contrib.auth.models import PermissionsMixin
+from django.contrib.auth.base_user import AbstractBaseUser
+from django.utils.translation import gettext_lazy as _
+from django.contrib.auth.base_user import BaseUserManager
+
+
+# Create your models here.
+class UserManager(BaseUserManager):
```
Comment (ksilawat-deqode) [inline EMS/users/models.py]: Convert it to private member for this particular module.

File: EMS/users/models.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+    def get_full_name(self):
+        '''
+        Returns the first_name plus the last_name, with a space in between.
+        '''
+        full_name = '%s %s' % (self.first_name, self.last_name)
+        return full_name.strip()
+
+    def __str__(self):
+        return self.email
```
Comment (ksilawat-deqode) [inline EMS/users/models.py]: No newline at the end.

**Review by ksilawat-deqode — CHANGES_REQUESTED** [review]
- Remove unused apps `Enquiry` & `user_app`
- Rename `EMS` folder to `ems` 
- Unapply all the migrations, delete them & recreate all migrations again(Only one migration should be there for each model as the definition of the models is cleared).

File: EMS/user_app/models.py (hunk 10 above)
Comment (ksilawat-deqode) [reply EMS/user_app/models.py]: Any answer to this question?

Comment (Rahul4321Kumar) [reply EMS/user_app/models.py]: ('M', 'Male'),
('F', 'Female'),
we also write like this 
first element of each tuple value is set to model and second element for human readable

File: EMS/users/models.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+
+class User(AbstractBaseUser, PermissionsMixin):
+    GENDER_CHOICES = (
+        (u'M', u'Male'),
+        (u'F', u'Female'),
+    )
+    email = models.EmailField(_('email address'), unique=True)
+    full_name = models.CharField(_('full name'), max_length=30, blank=True)
+    gender = models.CharField(_('gender'), max_length=2, choices=GENDER_CHOICES)
+    is_superuser = models.BooleanField(_('superuser'))
```
Comment (ksilawat-deqode) [inline EMS/users/models.py]: Please add a `default=False` for this field.

File: EMS/users/models.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
+from django.db import models
+from django.contrib.auth.models import PermissionsMixin
+from django.contrib.auth.base_user import AbstractBaseUser
+from django.utils.translation import gettext_lazy as _
```
Comment (ksilawat-deqode) [inline EMS/users/models.py]: Any reason for using this thing?

File: EMS/users/models.py [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+        if extra_fields.get('is_superuser') is not True:
+            raise ValueError('Superuser must have is_superuser=True.')
+
+        return self._create_user(email, password, **extra_fields)
+
+
+class User(AbstractBaseUser, PermissionsMixin):
+    GENDER_CHOICES = (
+        (u'M', u'Male'),
+        (u'F', u'Female'),
```
Comment (ksilawat-deqode) [inline EMS/users/models.py]: Still haven't got any reason, why you used this thing?

**Review by ksilawat-deqode — CHANGES_REQUESTED** [review]
Please do reply on question asked!

File: EMS/users/models.py (hunk 17 above)
Comment (Rahul4321Kumar) [reply EMS/users/models.py]: gettext_lazy is used to translate the string for end user its not neccessary for now

Comment (ksilawat-deqode) [reply EMS/users/models.py]: If it's not necessary for now, then remove it.

File: EMS/users/models.py (hunk 18 above)
Comment (Rahul4321Kumar) [reply EMS/users/models.py]: 'choices' must be an iterable containing (actual value, human readable name) tuples.

Comment (Rahul4321Kumar) [reply EMS/users/models.py]: in django-admin gender will be displayed as drop down

File: EMS/users/models.py [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    GENDER_CHOICES = (
+        ("M","Male"),
+        ("F","female"),
+    )
+    email = models.EmailField(unique=True)
+    full_name = models.CharField(max_length=30, blank=True)
+    gender = models.CharField(max_length=2, choices=GENDER_CHOICES)
+    is_superuser = models.BooleanField(default=False)
+    active = models.BooleanField(default=True)
+    is_staff = models.BooleanField(default=True)
```
Comment (ksilawat-deqode) [inline EMS/users/models.py]: This field should have `default=False`.
Do read about `is_staff` & `is_superuser` fields on `User` model.

**Review by ksilawat-deqode — CHANGES_REQUESTED** [review]

File: EMS/users/models.py (hunk 19 above)
Comment (Rahul4321Kumar) [reply EMS/users/models.py]: is_staff+is_superuser+active=True means its a main user(superuser)
is_staff+active=True means its a admin user
active=True mean its a normal user

File: EMS/users/models.py [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+        ("M","Male"),
+        ("F","female"),
+    )
+    email = models.EmailField(unique=True)
+    full_name = models.CharField(max_length=30, blank=True)
+    gender = models.CharField(max_length=2, choices=GENDER_CHOICES)
+    is_superuser = models.BooleanField(default=False)
+    active = models.BooleanField(default=True)
+    is_staff = models.BooleanField(default=False)
+    created_at = models.DateTimeField(auto_now_add = True)
```
Comment (ksilawat-deqode) [inline EMS/users/models.py]: Typo here
```suggestion
    created_at = models.DateTimeField(auto_now_add=True)
```

**Review by ksilawat-deqode — CHANGES_REQUESTED** [review]
Update ASAP, will merge it after the update.

File: EMS/users/models.py (hunk 18 above)
Comment (ksilawat-deqode) [reply EMS/users/models.py]: No, it's not an answer to the question asked

File: EMS/users/models.py (hunk 20 above)
Comment (Rahul4321Kumar) [reply EMS/users/models.py]: Done

**Review by kishansilawat — APPROVED** [review]
After this much iterations.

**Pull request merged**